LFS-1044: Newly added questions are not highlighted in the questionnaire editor

veronikaslc opened the pull request.

Commit:
LFS-1044: Newly added questions are not highlighted in the questionnaire editor

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/EditDialog.jsx:
```diff
@@ -96,13 +97,21 @@ let EditDialog = (props) => {
+        .then(() => {
+          setLastSaveStatus(true);
+          // Redirect to the created item to focus
+          props.history.push({
+            pathname: /Questionnaires\/([^#]+)/.exec(window.location.href)?.[1],
+            hash: `${data['@path']}/${targetId}`
+          });
```
Is there another way to highlight than by pushing a new URL to history? Now if the user opens a questionnaire and adds a new question, when clicking the browser back button they stay on the questionnaire page (but question loses highlight) and they have to click back again to actually navigate back.

Additionally, this highlights only the last question added and removes all existing highlighting, which is inconsistent with what happens on editing questions, where every questionnaire item modified in the current session remains highlighted.

sdumitriu replied:
I think there's a way to "replace" the history.

sdumitriu replied:
But edited existing questions don't need the hash, can't the same mechanism be used for new questions?

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/DeleteDialog.jsx:
```diff
@@ -57,7 +58,12 @@ let DeleteDialog = (props) => {
+        setOpen(false);
+        onClose && onClose();
+        // Redirect to the questionnaire
+        props.history.push(/Questionnaires\/([^#]+)/.exec(window.location.href)?.[1]);
```
This introduces the same issue with the back button and with highlighting removal as the creation of a new question.

Commit:
LFS-1044: Newly added questions are not highlighted in the questionnaire editor
Reloading newly created item and passing to the parent with loaded data to rerender and focus

veronikaslc commented:
@sashaandjic Ready for test

veronikaslc commented:
rebased on latest dev

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Questionnaire.jsx:
```diff
@@ -74,6 +74,11 @@ let Questionnaire = (props) => {
   }
 
+  let onCreate = (newData) => {
+    setData({});
```
What is the purpose of this line? Is it to force re-rendering?

veronikaslc replied:
yes

veronikaslc commented:
rebased on the latest dev again to resolve conflict

sashaandjic commented:
The newly added questions now deselect previously modified ones. Steps to reproduce:
- open any questionnaire
- change any 2 questions
- both questions will be highlighted
- add a new question
- the newly added question will be highlighted, but the two that were previously changed are no longer highlighted

Commit:
LFS-1065: Use fetchWithReLogin in the questionnaire wizard

veronikaslc commented:
@marta- The best way to address the bug is to close the PR untill the global context it implemented and completely reimplement 1044 then.

sashaandjic commented:
Removed bug label. It is good as is. The above mentioned problem is not an easy one to fix. The current code change offers a nice improvement.

sashaandjic reviewed (approved):
tested

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/EditDialog.jsx:
```diff
@@ -170,6 +172,7 @@ let EditDialog = (props) => {
           { dialogTitle() }
           </DialogTitle>
           <DialogContent>
+            { error && <Typography color="error">error</Typography>}
```
Did you mean this instead?

```suggestion
            { error && <Typography color="error">{error}</Typography>}
```

sdumitriu reviewed (commented):
It is finicky. Sometimes, but not always, adding a new question loses the highlight on (some) previously edited questions. Sometimes, but not always, cancelling the Add dialog reloads the questionnaire.

marta- committed:
LFS-1044: Newly added questions are not highlighted in the questionnaire editor

Fixed error display

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

veronikaslc committed:
Merge branch 'dev' into LFS-1044
